### PR TITLE
Ft user make request 168781684

### PR DESCRIPTION
**Pull request by Cheza-Dzabala**
#### What does this PR do?

This PR allows a user to place trip requests. 

#### Description of Task to be completed?

- Allow a user to place a one way trip request.
- Allow a user to place a return trip request.
- Allow a user to place a multi-city trip request.

#### How should this be manually tested?

1. Clone this repo
2. Run ` npm install `
3. Run ` sequelize db:migrate `
4. Run ` sequelize db:seed:all `
5. Run ` npm run dev `
6. Using POSTMAN, login is as a normal user using any method and retrieve the returned token.

    ###### Use the Bearer Auth Authentication hence forth

7. Structure the body as follows:
```javascript
{
        "typeId": INTEGER,
        "locationId": INTEGER,
        "departureDate": "YYYY-MM-DD",
        "returnDate": "YYYY-MM-DD",
        "destinations": [
            {
                "arrivalDate": "YYYY-MM-DD",
                "departureDate": "YYYY-MM-DD",
                "reasons": STRING,
                "isFinal": BOOLEAN,
                "locationId": INTEGER,
                "bookingId": INTEGER
            }
        ]
    }
```

___

##### Note:

typeId Values:

 1 = One way trip
2 = Return trip
3 = Multicity trip
___
8. Send a POST request to ` http://localhost:<:port>/api/v1/requests`

#### Any background context you want to provide?

N / A

#### Screenshots (if appropriate)

Screen Shot 2019-10-24 at 09 35 03
Screen Shot 2019-10-24 at 09 35 32
Screen Shot 2019-10-24 at 09 35 52
Screen Shot 2019-10-24 at 09 36 17
Screen Shot 2019-10-24 at 09 36 41

#### Questions:
N / A

### CODE REVIEW

**Review by alainmateso — CHANGES_REQUESTED** [review]
Great Job so far @Cheza-Dzabala . However, Please return data objects in your response where possible like the location's name at least. Make sure the problems of the departure date at the destination that can be greater than the return date are also fixed.

**Review by nignanthomas — CHANGES_REQUESTED** [review]
Impressive job with your validations @Cheza-Dzabala 👏 
Just implement the changes I told you on Slack

* tables foreignKeys should be >=1

* suppliers wich are external to the company should not be able to place requests.

File: src/controllers/destinationController.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import destinationServices from '../services/requestServices/destinationServices';
+import Utilities from '../utils/index';
+
+export default class destinationController {
+  static async storeDestination(res, body, request) {
+
```
Comment (nakiwuge) [inline src/controllers/destinationController.js:7]: Remove this line

File: src/controllers/requestController.js [hunk 2]
```diff
@@ -23,4 +25,10 @@ export default class requestController {
       200
     );
   }
+
```
Comment (nakiwuge) [inline src/controllers/requestController.js:102]: Remove this line

File: src/middlewares/requests/bookingsVerification.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  for (let counter = 0; counter < destinations.length; counter += 1) {
+    /*  eslint-disable-next-line max-len */
+    const bookingLength = destinations.filter(dest => dest.bookingId === destinations[counter].bookingId);
+    if (bookingLength.length > 1) {
+      return true;
+    }
+  }
+  return false;
+};
+
```
Comment (nakiwuge) [inline src/middlewares/requests/bookingsVerification.js:29]: Remove this line

File: src/middlewares/requests/bookingsVerification.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+const checkMultiple = destinations => {
```
Comment (nakiwuge) [inline src/middlewares/requests/bookingsVerification.js]: What are you trying to achieve with this function?

File: src/middlewares/requests/datesVerification.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
+import Utilities from '../../utils/index';
+import userServices from '../../services/userServices';
+
+const verify = destinations => {
+  for (let counter = 0; counter < destinations.length; counter += 1) {
```
Comment (nakiwuge) [inline src/middlewares/requests/datesVerification.js]: Why are you using a for loop here? What do you want to achieve?

File: src/middlewares/requests/datesVerification.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+const verifyTravelDates = async (departureDate, returnDate, destinations, user) => {
+  const query = Utilities.userQueries.userRequests(user.payload);
+  const { requests } = await userServices.findOne(query);
+
+  const result = {
+    dateVerificationError: false,
+    dateVerificationMessage: ''
+  };
+
+  for (let counter = 0; counter < requests.length; counter += 1) {
```
Comment (nakiwuge) [inline src/middlewares/requests/datesVerification.js]: Why are you using a for loop here? What do you want to achieve?

File: src/middlewares/requests/destinationVerification.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+const verify = destinations => {
+  let FinalFlag = false;
+  for (let counter = 0; counter < destinations.length; counter += 1) {
```
Comment (nakiwuge) [inline src/middlewares/requests/destinationVerification.js]: Why are you using a for loop here? What do you want to achieve?

File: src/middlewares/requests/locationVerification.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      }
+    }
+  );
+  const count = await requestServices.locations.countAll(locationQuery(id));
+  if (count === 0) {
+    return false;
+  }
+  return true;
+};
+
```
Comment (nakiwuge) [inline src/middlewares/requests/locationVerification.js:13]: Remove this line

File: src/middlewares/requests/locationVerification.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+const originalLocationCheck = (destinations, locationId) => {
+  for (let counter = 0; counter < destinations.length; counter += 1) {
```
Comment (nakiwuge) [inline src/middlewares/requests/locationVerification.js]: Why are you using a for loop here?

File: src/middlewares/requests/locationVerification.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    if (destinations[counter].locationId === locationId) {
+      return true;
+    }
+  }
+  return false;
+};
+
+
+const twoPointVerificationCheck = destinations => {
+  for (let counter = 0; counter < destinations.length; counter += 1) {
```
Comment (nakiwuge) [inline src/middlewares/requests/locationVerification.js]: Why are you using a for loop here?

File: src/middlewares/requests/relationVerification.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+import datesVerification from './datesVerification';
+import bookingsChecker from './bookingsVerification';
+import destinationChecker from './destinationVerification';
+import validationErrorFormatter from '../../utils/validationErrorFormatter';
+import Utilities from '../../utils/index';
+
+export default async (req, res, next) => {
+
+  const { body, user } = req;
+
```
Comment (nakiwuge) [inline src/middlewares/requests/relationVerification.js]: Remove this line

File: src/middlewares/requests/relationVerification.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+import locationChecker from './locationVerification';
+import tripTypesChecker from './tripTypesVerification';
+import datesVerification from './datesVerification';
+import bookingsChecker from './bookingsVerification';
+import destinationChecker from './destinationVerification';
+import validationErrorFormatter from '../../utils/validationErrorFormatter';
+import Utilities from '../../utils/index';
+
+export default async (req, res, next) => {
+
```
Comment (nakiwuge) [inline src/middlewares/requests/relationVerification.js]: Remove this line

File: src/middlewares/requests/relationVerification.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    return Utilities.responseHelper(
+      res,
+      Utilities.stringsHelper.validation.requests.locations.DOES_NOT_EXIST,
+      '',
+      400
+    );
+  }
+
+  /*  eslint-disable no-await-in-loop */
+  for (let counter = 0; counter < destinations.length; counter += 1) {
```
Comment (nakiwuge) [inline src/middlewares/requests/relationVerification.js]: Why are you using a for loop here?

**Review by alainmateso — APPROVED** [review]

**Review by nignanthomas — APPROVED** [review]

**Pull request merged**